File: src/gui/tile_proxy.py
"""
Local tile proxy — feeds Google Hybrid tiles to Qt's OSM plugin.

Qt's OSM plugin sends requests without browser headers, so Google
blocks them. This proxy runs on localhost, receives standard z/x/y
tile requests from Qt, and forwards them to Google with proper
Referer / User-Agent headers.  Includes disk cache for instant loads.
"""
import http.server
import logging
import os
import socket
import threading
import urllib.parse
import urllib.request

logger = logging.getLogger(__name__)

_GOOGLE_URL = "https://mt{s}.google.com/vt/lyrs=y&z={z}&x={x}&y={y}&scale={scale}"
_TILE_CACHE_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', '..', 'cache', 'tiles')

_HEADERS = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 '
                  '(KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
    'Referer': 'https://www.google.com/',
    'Accept': 'image/webp,image/apng,image/*,*/*;q=0.8',
}


class _TileHandler(http.server.BaseHTTPRequestHandler):
    """Handle /{z}/{x}/{y}[.ext] → Google Hybrid tile."""

    def do_GET(self):
        try:
            self._serve_tile()
        except (ConnectionAbortedError, ConnectionResetError, BrokenPipeError, OSError):
            pass  # Qt closed connection — normal during rapid panning

    def _serve_tile(self):
        parsed = urllib.parse.urlparse(self.path)
        path = parsed.path.strip('/')
        query = urllib.parse.parse_qs(parsed.query)
        # Strip file extension if Qt appends one (.png, .jpg, etc.)
        last = path.split('/')[-1] if '/' in path else path
        if '.' in last:
            path = path.rsplit('.', 1)[0]

        parts = path.split('/')
        if len(parts) != 3:
            self.send_error(404)
            return

        z, x, y = parts
        scale = 1

        # Qt high-DPI tile mode may request .../y@2x(.png) style paths.
        if y.endswith('@2x'):
            y = y[:-3]
            scale = 2

        # Optional explicit scale query (if client includes it).
        q_scale = query.get('scale')
        if q_scale:
            try:
                if int(q_scale[0]) >= 2:
                    scale = 2
            except (ValueError, TypeError):
                pass

        # ── Disk cache ──
        cache_path = os.path.join(_TILE_CACHE_DIR, z, f"{x}_{y}_s{scale}.tile")
        if os.path.exists(cache_path):
            try:
                with open(cache_path, 'rb') as f:
                    tile = f.read()
                self.send_response(200)
                self.send_header('Content-Type', 'image/png')
                self.send_header('Content-Length', str(len(tile)))
                self.send_header('Cache-Control', 'max-age=604800')
                self.end_headers()
                self.wfile.write(tile)
                return
            except Exception:
                pass  # cache read failed, fall through to fetch

        # ── Fetch from Google ──
        server = int(x) % 4
        url = _GOOGLE_URL.format(s=server, z=z, x=x, y=y, scale=scale)

        req = urllib.request.Request(url, headers=_HEADERS)
        try:
            with urllib.request.urlopen(req, timeout=10) as resp:
                tile = resp.read()
                ctype = resp.headers.get('Content-Type', 'image/png')
        except Exception as e:
            logger.debug("[TileProxy] fetch failed z=%s x=%s y=%s: %s", z, x, y, e)
            self.send_error(502)
            return

        # ── Save to disk cache ──
        try:
            os.makedirs(os.path.dirname(cache_path), exist_ok=True)
            with open(cache_path, 'wb') as f:
                f.write(tile)
        except Exception:
            pass

        self.send_response(200)
        self.send_header('Content-Type', ctype)
        self.send_header('Content-Length', str(len(tile)))
        self.send_header('Cache-Control', 'max-age=604800')
        self.end_headers()
        self.wfile.write(tile)

    def log_message(self, fmt, *args):
        pass  # suppress per-request HTTP logs


def start_tile_proxy() -> int:
    """Start tile proxy on a free port. Returns port number."""
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
        s.bind(('127.0.0.1', 0))
        port = s.getsockname()[1]

    server = http.server.ThreadingHTTPServer(('127.0.0.1', port), _TileHandler)
    t = threading.Thread(target=server.serve_forever, daemon=True)
    t.start()
    logger.info("[TileProxy] Google Hybrid proxy on http://127.0.0.1:%d", port)
    return port

File: src/gui/TileProxy.cpp
#include "TileProxy.h"
#include <QTcpSocket>
#include <QNetworkRequest>
#include <QNetworkReply>
#include <QFile>
#include <QDir>
#include <QUrl>
#include <QPointer>
#include <QCoreApplication>
#include <spdlog/spdlog.h>

namespace vtol {

TileProxy::TileProxy(QObject* parent)
    : QObject(parent)
{
    connect(&m_server, &QTcpServer::newConnection, this, &TileProxy::onNewConnection);
}

int TileProxy::startProxy()
{
    if (!m_server.listen(QHostAddress::LocalHost, 0)) {
        SPDLOG_ERROR("[TileProxy] Failed to listen: {}", m_server.errorString().toStdString());
        return -1;
    }
    m_port = m_server.serverPort();
    SPDLOG_INFO("[TileProxy] Google Hybrid proxy on http://127.0.0.1:{}", m_port);
    return m_port;
}

QString TileProxy::baseUrl() const
{
    return QString("http://127.0.0.1:%1").arg(m_port);
}

void TileProxy::onNewConnection()
{
    while (auto* socket = m_server.nextPendingConnection()) {
        // Prevent double-processing: disconnect readyRead after first complete read
        auto* conn = new QMetaObject::Connection;
        *conn = connect(socket, &QTcpSocket::readyRead, this, [this, socket, conn]() {
            disconnect(*conn);
            delete conn;
            QByteArray data = socket->readAll();
            handleRequest(socket, data);
        });
        connect(socket, &QTcpSocket::disconnected, socket, &QTcpSocket::deleteLater);
    }
}

void TileProxy::handleRequest(QTcpSocket* socket, const QByteArray& requestData)
{
    QString request = QString::fromUtf8(requestData);
    QStringList lines = request.split("\r\n");
    if (lines.isEmpty()) {
        sendError(socket, 400, "Bad Request");
        return;
    }

    QStringList parts = lines[0].split(' ');
    if (parts.size() < 2 || parts[0] != "GET") {
        sendError(socket, 400, "Bad Request");
        return;
    }

    QString path = parts[1];
    int qIdx = path.indexOf('?');
    QString query;
    if (qIdx >= 0) {
        query = path.mid(qIdx + 1);
        path = path.left(qIdx);
    }

    path = path.trimmed();
    while (path.startsWith('/')) path = path.mid(1);
    while (path.endsWith('/')) path.chop(1);

    QStringList segments = path.split('/');
    if (!segments.isEmpty()) {
        QString& last = segments.last();
        int dotIdx = last.lastIndexOf('.');
        if (dotIdx > 0) last = last.left(dotIdx);
    }

    if (segments.size() != 3) {
        sendError(socket, 404, "Not Found");
        return;
    }

    bool ok1, ok2, ok3;
    int z = segments[0].toInt(&ok1);
    int x = segments[1].toInt(&ok2);

    QString yStr = segments[2];
    int scale = 1;
    if (yStr.endsWith("@2x")) {
        yStr.chop(3);
        scale = 2;
    }
    int y = yStr.toInt(&ok3);

    if (!ok1 || !ok2 || !ok3) {
        sendError(socket, 404, "Not Found");
        return;
    }

    if (query.contains("scale=2")) scale = 2;

    serveTile(socket, z, x, y, scale);
}

void TileProxy::serveTile(QTcpSocket* socket, int z, int x, int y, int scale)
{
    // Check disk cache first
    QString cPath = cachePath(z, x, y, scale);
    QFile cached(cPath);
    if (cached.exists() && cached.open(QIODevice::ReadOnly)) {
        QByteArray tile = cached.readAll();
        cached.close();
        if (tile.size() > 100) {
            QByteArray ct = "image/jpeg";
            if (tile.size() > 4 && tile[0] == '\x89' && tile[1] == 'P')
                ct = "image/png";
            sendTileResponse(socket, tile, ct);
            return;
        }
        QFile::remove(cPath);
    }

    // Throttle concurrent fetches to avoid UI freezes during rapid zoom
    if (m_pendingFetches >= MAX_CONCURRENT_FETCHES) {
        m_fetchQueue.append({QPointer<QTcpSocket>(socket), z, x, y, scale});
        return;
    }

    // Fetch from Google — use QPointer to safely track socket lifetime
    QPointer<QTcpSocket> safeSocket(socket);
    m_pendingFetches++;

    QUrl url(googleUrl(z, x, y, scale));
    QNetworkRequest req(url);
    req.setRawHeader("User-Agent",
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
        "(KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36");
    req.setRawHeader("Referer", "https://www.google.com/");
    req.setRawHeader("Accept", "image/avif,image/apng,image/png,image/jpeg,*/*;q=0.8");

    QNetworkReply* reply = m_nam.get(req);

    connect(reply, &QNetworkReply::finished, this, [this, reply, safeSocket, z, x, y, scale]() {
        reply->deleteLater();
        m_pendingFetches--;

        // Socket destroyed while we were fetching? Just cache the tile.
        QByteArray rawTile;
        bool socketAlive = safeSocket && safeSocket->isOpen();

        if (reply->error() != QNetworkReply::NoError) {
            SPDLOG_WARN("[TileProxy] fetch failed z={} x={} y={}: {}",
                         z, x, y, reply->errorString().toStdString());
            if (socketAlive) sendError(safeSocket, 502, "Bad Gateway");
            processNextQueued();
            return;
        }

        rawTile = reply->readAll();

        if (rawTile.size() < 100) {
            SPDLOG_WARN("[TileProxy] Too small response: size={}", rawTile.size());
            if (socketAlive) sendError(safeSocket, 502, "Bad Gateway");
            processNextQueued();
            return;
        }

        QString ct = reply->header(QNetworkRequest::ContentTypeHeader).toString();
        if (m_loggedRequests < 3) {
            SPDLOG_INFO("[TileProxy] Tile z={} x={} y={}: {}B ct={}",
                         z, x, y, rawTile.size(), ct.toStdString());
            m_loggedRequests++;
        }

        // Save to disk cache
        QString cPath = cachePath(z, x, y, scale);
        QDir().mkpath(QFileInfo(cPath).path());
        QFile f(cPath);
        if (f.open(QIODevice::WriteOnly)) {
            f.write(rawTile);
            f.close();
        }

        if (socketAlive)
            sendTileResponse(safeSocket, rawTile, ct.toUtf8());

        processNextQueued();
    });
}

void TileProxy::processNextQueued()
{
    while (!m_fetchQueue.isEmpty() && m_pendingFetches < MAX_CONCURRENT_FETCHES) {
        auto req = m_fetchQueue.takeFirst();
        if (req.socket && req.socket->isOpen())
            serveTile(req.socket, req.z, req.x, req.y, req.scale);
    }
}

void TileProxy::sendTileResponse(QTcpSocket* socket, const QByteArray& tile, const QByteArray& contentType)
{
    QByteArray response;
    response.append("HTTP/1.1 200 OK\r\n");
    response.append("Content-Type: " + contentType + "\r\n");
    response.append("Content-Length: " + QByteArray::number(tile.size()) + "\r\n");
    response.append("Cache-Control: max-age=604800\r\n");
    response.append("Access-Control-Allow-Origin: *\r\n");
    response.append("Connection: close\r\n");
    response.append("\r\n");
    response.append(tile);

    socket->write(response);
    socket->flush();
    // Let the client close first (Connection: close), or auto-close after timeout
    connect(socket, &QTcpSocket::bytesWritten, socket, [socket]() {
        if (socket->bytesToWrite() == 0)
            socket->disconnectFromHost();
    });
}

void TileProxy::sendError(QTcpSocket* socket, int code, const QByteArray& reason)
{
    QByteArray response;
    response.append("HTTP/1.1 " + QByteArray::number(code) + " " + reason + "\r\n");
    response.append("Content-Length: 0\r\n");
    response.append("Connection: close\r\n");
    response.append("\r\n");

    socket->write(response);
    socket->flush();
    socket->disconnectFromHost();
}

QString TileProxy::cachePath(int z, int x, int y, int scale)
{
    return QCoreApplication::applicationDirPath() + "/" + CACHE_DIR
           + "/" + QString::number(z) + "/" + QString::number(x)
           + "_" + QString::number(y) + "_s" + QString::number(scale) + ".tile";
}

QString TileProxy::googleUrl(int z, int x, int y, int scale)
{
    int server = x % 4;
    return QString("https://mt%1.google.com/vt/lyrs=y&z=%2&x=%3&y=%4&scale=%5")
           .arg(server).arg(z).arg(x).arg(y).arg(scale);
}

} // namespace vtol
